File: 2_knapsack/solver.cpp
#include <vector>
#include <string>
#include <iostream>
#include <sstream>

std::string solver(std::string &input_data) {
  std::stringstream input_stream(input_data);

  int item_count, capacity;
  input_stream >> item_count >> capacity;

  std::vector<int> values(item_count);
  std::vector<int> weights(item_count);
  for (size_t i = 0; i < item_count; i++) {
    input_stream >> values[i] >> weights[i];
  }

  int total_value = 0;
  std::vector<int> taken(item_count);
  for (size_t i = 0; i < item_count; i++) {
    if (weights[i] <= capacity) {
      total_value += values[i];
      capacity -= weights[i];
      taken[i] = 1;
    } else {
      taken[i] = 0;
    }
  }

  std::stringstream solution;
  solution << total_value << " " << 0 << "\n";
  if (item_count > 0) {
    for (size_t i = 0; i < item_count - 1; i++) {
      solution << taken[i] << " ";
    }
    solution << taken[item_count - 1] << "\n";
  }
  return solution.str();
}

int main() {
  std::string input_data = "4 11\n8 4\n10 5\n15 8\n4 3\n";
  std::cout << solver(input_data);
  return 0;
}

File: 2_knapsack/solver.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

import sys
import methods


def solve_it(input_data):
    lines = input_data.split('\n')
    firstLine = lines[0].split()
    item_count = int(firstLine[0])
    capacity = int(firstLine[1])
    values = [None for _ in range(item_count)]
    weights = [None for _ in range(item_count)]
    for i in range(item_count):
        line = lines[i+1].split()
        values[i] = int(line[0])
        weights[i] = int(line[1])
    capacity, weights = methods.scale_inputs(capacity, weights, item_count)
    result = methods.branch_and_bound(values, weights, capacity)
    return result


if __name__ == '__main__':
    if len(sys.argv) > 1:
        file_location = sys.argv[1].strip()
        with open(file_location, 'r') as input_data_file:
            input_data = input_data_file.read()
        print(solve_it(input_data))
    else:
        print('This test requires an input file.  Please select one from the data directory. (i.e. python solver.py ./data/ks_4_0)')
